### ISSUE TITLE
Concern about partial functions in Conv

### ISSUE THREAD

**Issue by seagreen** [sdiehl/protolude#35]
While `StringConv` makes perfect sense in some instances (such as `StringConv String Text`), conversions from bytes to strings is an operation that can fail. However `StringConv` doesn't produce a result that includes failures.

Because of this instances like`StringConv ByteString Text` can either throw runtime errors or produce incorrect data, depending on whether `toS` or `toSL` is used.

Have you considered restricting `StringConv` to only conversions that can be made without errors?

**Comment by sdiehl** [comment]
In 0.2 I'll be exporting the ``Encoding`` and ``decodeUtf8'`` functions so users can manually specify if they want to handle Unicode failures:

https://github.com/sdiehl/protolude/blob/master/src/Protolude.hs#L360

**Comment by seagreen** [comment]
That helps some, but I feel like `StringConv` will remain a footgun even if Protolude also exports ways to do decoding correctly.

Wouldn't it be better to remove the `StringConv` instances that lead to partial functions entirely? I think it's only six:

```
StringConv B.ByteString String
StringConv B.ByteString T.Text
StringConv B.ByteString LT.Text

StringConv LB.ByteString String
StringConv LB.ByteString T.Text
StringConv LB.ByteString LT.Text
```

Perhaps a new class could be introduced to get some of the convenience of these instances back, but with safety. Something like:

```haskell
class StringEncoding a b where
     decodeString :: a -> Either UnicodeException b

instance StringEncoding B.ByteString T.Text where
    decodeString = decodeUtf8'
```

Since Protolude does such a great job getting rid of partial functions from the Prelude, it seems a shame to let new ones slip in due to this issue.

**Comment by sdiehl** [comment]
This seems like a sensible proposal, let me think about it and find out much code it might break. 

The next version will introduce breaking changes with a major version bump anyways so now might be the best time.

**Comment by seagreen** [comment]
Sounds great! Btw I'm not attached to the `StringEncoding` part of my proposal at all, so if you have a better way of allowing non-partial decoding please go with it.

**Comment by seagreen** [comment]
By the way, /r/haskell just pointed me to [text-conversions](https://github.com/cjdev/text-conversions) which handles this stuff safely and might be a useful inspiration.

**Comment by PierreR** [comment]
What is the status on this issue now that protolude-0.2 is out ?

**Comment by sdiehl** [comment]
The plan will be to provide two typeclasses, one of which is partial and behaves like `toS` today. And then create `MaybeStringConv` which forces the user to the handle the `Stringconv a b => a -> Maybe b` explicitly. This will be in the 0.2 version which will be cut when the Stackage LTS rolls over to GHC 8.2.

**Comment by PierreR** [comment]
Thanks for the quick update. By the way I thought `0.2` was already released (it is on hackage but not on github)

**Comment by seagreen** [comment]
Is the plan to eventually deprecate the partial one, or is having two typeclasses the long term plan?

**Comment by sdiehl** [comment]
We could on the next major version bump. Although it seems like string conversion is one of the few cases where people want the partiality (in favor of handling Eithers) because of how frequently it occurs.

**Comment by seagreen** [comment]
> Although it seems like string conversion is one of the few cases where people want the partiality (in favor of handling Eithers) because of how frequently it occurs.

I agree people feel this way, all the partial functions in [Data.Text.Encoding](https://hackage.haskell.org/package/text-1.2.2.2/docs/Data-Text-Encoding.html) provide evidence for this.

Unfortunately the main driver of my interest in alternate Preludes is to get away from unnecessary partial functions. So my preference would be to deprecate the typeclass no matter how popular it may be.**

Another option would be to make a `Protolude.Scripting` or `Protolude.Partial` module for situations where users don't case so much about runtime errors. Partial functions are genuinely convenient for quick-and-dirty scripts that don't do any error handling, but crash on the first unexpected `Left`. `toS` might be a good fit there. I'm not sure this is a good idea, but at least it's an option.

** The more I think about it, the more I think that the fact people are tempted to cheat and use partial functions for string conversions is a good reason not to encourage them by providing those functions in a prelude. It makes me think of [this article](https://pchiusano.github.io/2017-01-20/why-not-haskell.html) about how Haskell doesn't really shine at serialization/IO boundaries. That may be true, but there's no reason not to at least take them as seriously as we can.

**Comment by sdiehl** [comment]
@seagreen You have commit now so I'm cool considering whatever interface you want for string conversions now. I'm in favor of the two module approach.

**Comment by seagreen** [comment]
Sounds great, I'll do some thinking about it and make a concrete proposal.

**Comment by seagreen** [comment]
Did some thinking on the situation and put up a draft PR here: https://github.com/protolude/protolude/pull/113

My thoughts:

1. In the spirit of not defaulting users to partial functions, I removed the `Protolude.Conv` re-export from `Protolude`. I haven't touched `Protolude.Conv` otherwise except to add a header doc. We might want to rename it though to reflect its partiality. Or we could fold it into a new `Protolude.Scripting` module where we could put all the unsafe stuff that isn't re-exported by default.

2. Due the the very annoying `String` situation in Haskell, a really important use case for me is a lightweight way to convert between `Text`/`TL.Text`/`String`. Conversions between those three are safe, so `MaybeStringConv` is overkill in that case. Instead I added a typeclass:
    ```hs
    -- | Convert from one Unicode textual type to another. Not for serialization/deserialization,
    -- so doesn't have instances for bytestrings.
    class ConvertText a b where
      convertText :: a -> b
    ```

Some questions:

1. Should we add a `MaybeStringConv` typeclass as well?

2. While `ConvertText` can't support converting from bytestrings, it could support converting to them if we're willing to assume UTF-8, should we do that? Currently I have two helper functions `toUtf8` and `toUtf8Lazy` to help with that, but we build it right into the typeclass.

My personal thinking is that I like the conceptual simplicity of `ConverText`-- that it's just for Unicode conversions, not serializing. However, I'm definitely open to allowing Utf-8 serializing if you want.
